### PR TITLE
fix(kuma-cp): treat envoy admin errors as 4xx

### PR DESCRIPTION
**Pull request by lobkovilya**
### Checklist prior to review

- [X] [Link to relevant issue][1] as well as docs and UI issues --
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/envoy/admin/kds_client.go [hunk 1]
```diff
@@ -163,3 +165,20 @@ func resNameInZone(nameInGlobal string, k8sStore bool) (string, string, error) {
...
+func (e *KDSTransportError) Error() string {
+	if e.reason == "" {
+		return fmt.Sprintf("could not send %s", e.requestType)
+	} else {
+		return fmt.Sprintf("could not send %s: %s", e.requestType, e.reason)
+	}
+}
+
+func (e *KDSTransportError) Is(err error) bool {
+	return reflect.TypeOf(e) == reflect.TypeOf(err)
```
Comment (michaelbeaumont) [inline pkg/envoy/admin/kds_client.go:183]: Shouldn't this just be like:

```suggestion
	_, ok := target.(*KDSTransportError)
	return ok
```

?

**Review by michaelbeaumont — APPROVED** [review]

File: pkg/envoy/admin/kds_client.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/envoy/admin/kds_client.go:183]: All other `Is` implementations are using `reflect`:
```
$ grep -r ") Is(err" .
./pkg/api-server/types/errors.go:func (a *InvalidPageSizeError) Is(err error) bool {
./pkg/core/access/errors.go:func (a *AccessDeniedError) Is(err error) bool {
./pkg/core/resources/store/store.go:func (e *ResourceConflictError) Is(err error) bool {
./pkg/core/resources/store/store.go:func (e *AssertionError) Is(err error) bool {
./pkg/core/resources/store/store.go:func (a *PreconditionError) Is(err error) bool {
./pkg/core/rest/errors/errors.go:func (e *BadRequest) Is(err error) bool {
./pkg/intercp/envoyadmin/forwarding_kds_client.go:func (e *StreamNotConnectedError) Is(err error) bool {

$ grep -r "reflect.TypeOf(.*) == reflect.TypeOf(err)" .
./pkg/api-server/types/errors.go:       return reflect.TypeOf(a) == reflect.TypeOf(err)
./pkg/core/access/errors.go:    return reflect.TypeOf(a) == reflect.TypeOf(err)
./pkg/core/resources/store/store.go:    return reflect.TypeOf(e) == reflect.TypeOf(err)
./pkg/core/resources/store/store.go:    return reflect.TypeOf(e) == reflect.TypeOf(err)
./pkg/core/resources/store/store.go:    return reflect.TypeOf(a) == reflect.TypeOf(err)
./pkg/core/rest/errors/errors.go:       return reflect.TypeOf(e) == reflect.TypeOf(err)
./pkg/intercp/envoyadmin/forwarding_kds_client.go:      return reflect.TypeOf(e) == reflect.TypeOf(err)

```

If we want to change this I can open another PR that updates all the places

Comment (lahabana) [reply pkg/envoy/admin/kds_client.go:183]: Is there much a difference?

File: pkg/core/rest/errors/error_handler.go [hunk 2]
```diff
@@ -145,6 +146,12 @@ func HandleError(ctx context.Context, response *restful.Response, err error, tit
 			Title:  title,
 			Detail: err.Error(),
 		}
+	case errors.Is(err, &admin.KDSTransportError{}):
+		kumaErr = &types.Error{
+			Status: 400,
```
Comment (lahabana) [inline pkg/core/rest/errors/error_handler.go:151]: Is 400 really the right status here? There's no user error afaiu. I think if we want to use 4xx we should use another status code (potentially a non standard one)

**Comment by lahabana** [comment]
@lobkovilya what's missing to get this in?

File: pkg/core/rest/errors/error_handler.go (hunk 2 above)
Comment (jakubdyszkiewicz) [reply pkg/core/rest/errors/error_handler.go:151]: I'm fine either with 400 or with non-standard 4xx error. It both indicates user-facing error which is technically not correct

Comment (lobkovilya) [reply pkg/core/rest/errors/error_handler.go:151]: It's tricky to pick a good status code here, but I was thinking "who can do something to resolve the problem" and in this case, it's a user. So it's a bit of a stretch, but I think it's acceptable to treat it as a user error

File: pkg/envoy/admin/kds_client.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/envoy/admin/kds_client.go:183]: Avoiding reflections (if possible) is a good approach in general. But I don't think we should change it in this PR.

File: pkg/core/rest/errors/error_handler.go (hunk 2 above)
Comment (michaelbeaumont) [reply pkg/core/rest/errors/error_handler.go:151]: Do we also need to consider how the GUI responds to this error?

Comment (lobkovilya) [reply pkg/core/rest/errors/error_handler.go:151]: Not really, it's just going to display the error text, we already have a similar 400 error for StreamNotConnected https://github.com/kumahq/kuma/blob/9146696f3a05a8c0bc70c0b711a1b1ac8e75914e/pkg/core/rest/errors/error_handler.go#L131

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**